Upgrade fluentd-elasticsearch addon to Elasticsearch/Kibana 5.5

aknuds1 opened the pull request:
This is a patch to upgrade the fluentd-elasticsearch addon to Elasticsearch/Kibana 5.5. Please provide feedback!

```release-notes
* Upgrade Elasticsearch/Kibana to 5.5.1 in fluentd-elasticsearch addon
* Switch to basing our image of Elasticsearch in fluentd-elasticsearch addon off the official one
* Switch to the official image of Kibana in fluentd-elasticsearch addon
* Use StatefulSet for Elasticsearch instead of ReplicationController, with persistent volume claims
* Require authenticating towards Elasticsearch, as Elasticsearch 5.5 by default requires basic authentication
```

aknuds1 committed:
Upgrade Elasticsearch/Kibana to 5.5.1 and use official Kibana image

coffeepac commented:
@aknuds thanks for the PR!  Do you want any feedback on this now?   There are a few things that will prevent me from kicking off tests.

aknuds1 commented on cluster/addons/fluentd-elasticsearch/es-image/run.sh:
```diff
@@ -1,30 +1,11 @@
-chown -R elasticsearch:elasticsearch /data
-
-exec gosu elasticsearch sh /elasticsearch/bin/elasticsearch
+sysctl -w vm.max_map_count=262144
```
This seems to cause a warning about the filesystem being read only. Is it even needed?

coffeepac replied:
I'm not sure if its required for the es5 image, it may already be set.  however, ES5 has a strictly enforced minimum number of memory maps that must be available or ES refuses to start.

aknuds1 replied:
@coffeepac The official ES5 image does not call `sysctl`, nor does it seem to set `vm.mx_map_count` in any other way.

coffeepac replied:
I believe this is still enforced by ES 5.5:  https://www.elastic.co/guide/en/elasticsearch/reference/5.x/_maximum_map_count_check.html

there is another workaround but I don't fully understand it:
https://www.elastic.co/guide/en/elasticsearch/reference/5.x/breaking-changes-5.5.html#breaking_55_packaging_changes

aknuds1 commented on cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml:
```diff
@@ -24,14 +24,19 @@ spec:
         env:
         - name: FLUENTD_AGRS
           value: -q
+        # TODO: Make into secret!!
```
What do we do about these Elasticsearch credentials? Should we make them configurable via a Makefile?

coffeepac replied:
what are these used for?  is this basic auth against the ES cluster?

aknuds1 replied:
@coffeepac That is correct, since authentication is enabled by default for ES 5.5.

aknuds1 commented:
@coffeepac I would love any feedback at this point! I'm really especially interested in hearing about my approach when it comes to the Elasticsearch Dockerfile. I've justed based it off the original one, instead of writing it from scratch, as was done in the original.

aknuds1 commented on cluster/addons/fluentd-elasticsearch/es-image/Dockerfile:
```diff
@@ -11,44 +11,46 @@
+USER elasticsearch
+# Temporarily move config out of the way before installing plugin as installation will fail
+# otherwise
+RUN mv config config.bak && ./bin/elasticsearch-plugin install -b repository-s3 && \
```
I suspect that which plugins to install, repository-s3 in this case, should be configurable in some way.

coffeepac replied:
yes and I'm not sure how it could be done.

aknuds1 replied:
@coffeepac I'm sure there could be a way. I might have seen it being done in another repository in fact.

coffeepac replied:
if you have seen it I would love to get that in here.  there are several things that should be parameterized.

coffeepac commented on cluster/addons/fluentd-elasticsearch/es-image/run.sh:
```diff
@@ -1,30 +1,11 @@
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
```
this copyright information should remain and should be at the top of all files

coffeepac commented on cluster/addons/fluentd-elasticsearch/es-image/run.sh:
```diff
@@ -1,30 +1,11 @@
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
-sysctl -w vm.max_map_count=262144
```
does the elasticsearch 5 image handle this?  Its a requirement for ES5 to start.

coffeepac commented on cluster/addons/fluentd-elasticsearch/es-statefulset.yaml:
```diff
@@ -22,7 +24,7 @@ spec:
       serviceAccountName: elasticsearch-logging
       containers:
-      - image: gcr.io/google_containers/elasticsearch:v2.4.1-2
+      - image: gcr.io/google_containers/elasticsearch-k8s-s3:5.5.0-11
```
having s3 included in the default is a bit tough as mentioned above.  Its not a useful plugin for GKE clusters.

coffeepac commented on cluster/addons/fluentd-elasticsearch/es-statefulset.yaml:
```diff
@@ -38,13 +40,19 @@ spec:
+  - metadata:
+      name: elasticsearch-logging
+    spec:
+      storageClassName: gp2
```
this isn't portable, its AWS specific. allow this to be the cluster default (which I think is an empty string?  the semantics of storage class default is very particular).

coffeepac commented on cluster/addons/fluentd-elasticsearch/es-statefulset.yaml:
```diff
@@ -38,13 +40,19 @@ spec:
+      accessModes: ["ReadWriteOnce"]
+      resources:
+        requests:
+          storage: 20Gi
```
yes, 20Gi is far too low for a production use case.

aknuds1 replied:
@coffeepac How should we make it configurable in your opinion?

coffeepac commented on cluster/addons/fluentd-elasticsearch/fluentd-es-ds.yaml:
```diff
@@ -1,20 +1,20 @@
     kubernetes.io/cluster-service: "true"
     addonmanager.kubernetes.io/mode: Reconcile
-    version: v1.22
+    version: v1.23
```
I think the current version is 24?  probably shouldn't change it yet as there may be other PRs that bump the version but something to check when closer to merge.

coffeepac commented:
I don't have any ideas for how to provide a list of plugins to be installed at install time.  @crassirostris do you have any thoughts on that?  its something I've been considering but I don't know if this is something that there is already a pattern for elsewhere in kubernetes.

coffeepac commented:
/ok-to-test

aknuds1 commented:
@coffeepac I will give the PR another look today, I have some things in mind that I could do in order to solve the remaining problems.

aknuds1 commented:
I tested the PR now and I'm encountering a bug on the startup of the pod. Really note sure what's causing this as I've not encountered it in prior testing :( Gonna have to try some more tomorrow.

    [2017-07-14T21:11:06,097][INFO ][o.e.n.Node               ] [elasticsearch-logging-v1-0] initializing ...
    [2017-07-14T21:11:06,174][WARN ][o.e.b.ElasticsearchUncaughtExceptionHandler] [elasticsearch-logging-v1-0] uncaught exception in thread [main]
    org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: Failed to create node environment
    	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:127) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:114) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:67) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:122) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.cli.Command.main(Command.java:88) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:91) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:84) ~[elasticsearch-5.5.0.jar:5.5.0]
    Caused by: java.lang.IllegalStateException: Failed to create node environment
    	at org.elasticsearch.node.Node.<init>(Node.java:267) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.node.Node.<init>(Node.java:244) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:232) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:232) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:351) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:123) ~[elasticsearch-5.5.0.jar:5.5.0]
    	... 6 more
    Caused by: java.nio.file.AccessDeniedException: /data/nodes
    	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:84) ~[?:?]
    	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[?:?]
    	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[?:?]
    	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:384) ~[?:?]
    	at java.nio.file.Files.createDirectory(Files.java:674) ~[?:1.8.0_131]
    	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:781) ~[?:1.8.0_131]
    	at java.nio.file.Files.createDirectories(Files.java:767) ~[?:1.8.0_131]
    	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:221) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.node.Node.<init>(Node.java:264) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.node.Node.<init>(Node.java:244) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:232) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:232) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:351) ~[elasticsearch-5.5.0.jar:5.5.0]
    	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:123) ~[elasticsearch-5.5.0.jar:5.5.0]
    	... 6 more

coffeepac commented:
@aknuds1 looks like the /data mount point is being created with permissions that the `elasticsearch` user can't use.  maybe try removing the GOSU pieces for now and let ES run as root.

aknuds1 commented:
@coffeepac I made a mistake when I tested last night, it actually works :)

I've tested the whole stack now - I can see log messages in Kibana, so I think we can work towards merging it now! :)

aknuds1 commented:
@coffeepac Do you know how to fix whatever's causing the tests to break?

crassirostris commented:
@aknuds1 You somehow broke the licence boilerplate in `cluster/addons/fluentd-elasticsearch/es-image/run.sh` and `cluster/addons/fluentd-elasticsearch/es-image/Dockerfile`. Run `hack/verify-boilerplate.sh` from the K8s root directory to get that error. I guess there should be a newline between licence and the content.

crassirostris commented:
@aknuds1 I have several comments:

- You don't need `-1` in the end of the image tag in the Makefile, it was added to have two images with the same version of ES and different base images (for fixing CVEs)
- Don't run it as root in the end
- Do you really need to install gosu from Github? Maybe you can roll without it somehow?
- You removed the Kibana image and `KIBANA_BASE_URL` parameter. Have you verified that Kibana works through proxy?

aknuds1 commented:
@crassirostris Thanks for the comments! 

* I removed the -1 part
* I removed the need for gosu by running as elasticsearch in the end
* I removed the Kibana image as from what I can tell we can use the official one. I haven't yet tested with proxy

crassirostris commented:
@aknuds1 Cool, thanks!

> I removed the Kibana image as from what I can tell we can use the official one. I haven't yet tested with proxy

Yeah, I understand why, thanks for that :) Just want to make sure it works both when accessed directly and through proxy

aknuds1 commented:
@crassirostris I tried now, it doesn't work through proxy. The reason I removed KIBANA_BASE_URL however, is with this setting `kubectl port-forward` was broken. I use `kubectl port-forward` in order to access Kibana within the cluster, instead of via the `kubectl proxy`. I wish I could make both of them work, but I haven't found any configuration that works for them both so far, and I would rather go with `kubectl port-forward` since it's easier.

coffeepac commented:
@crassirostris there are a lot of somewhat surprising (to me) testing failures.  are these actually part of this PR or are other things breaking and this PR is just hitting them?  

ie:  I have no idea why this PR would cause cross to fail with some docker windows type mis-match failure.

aknuds1 commented:
I tried to fix the boilerplate issues.

crassirostris commented:
Somebody broke it everywhere: https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-cross

crassirostris commented:
Related issue: https://github.com/kubernetes/kubernetes/issues/48887

crassirostris commented:
@aknuds1 

> I would rather go with kubectl port-forward since it's easier.

OK, but then it definitely should be noted in the release notes & there should be a mention in the README somewhere. Also it'll break e2e test (`test/e2e/instrumentation/logging/es_kibana.go`), I'd be really grateful it you could adjust them (in a separate PR), I'm ready to help with that

aknuds1 commented:
@crassirostris Sure, I can fix those matters. Do you agree with supporting `kubectl port-forward` rather than `kubectl proxy`? I'm not sure about what's best, I'm just used to the port forward method myself and find it simpler than going via the API proxy.

crassirostris commented:
@aknuds1 Tbh, I don't care, but I also find binding to root more sensible. Let's just provide a better experience for those who don't know about this... peculiarity

aknuds1 commented:
@crassirostris That's my take too, it was pretty tricky for me to find out that I needed to go via the proxy and what the corresponding URL would be :)

aknuds1 commented:
@crassirostris I created [a separate PR](https://github.com/kubernetes/kubernetes/pull/49060) as your requested for the KIBANA_BASE_URL modification.

aknuds1 commented:
@crassirostris @coffeepac I decided to use gosu again to execute ES as the elasticsearch user, for the reason that I found today that /data or its contents might not be owned by elasticsearch on runtime, so that the run.sh script needs to run as root and chown /data. Hope you're OK with this, I've found it solves my problem.

crassirostris commented:
@aknuds1 OK, but could you please try to use `chroot --userspec=USER:GROUP` instead? I don't like installing something from github. @coffeepac Do you agree?

aknuds1 commented:
@crassirostris Should it be `chroot --userspec=elasticsearch:elasticsearch /`?

crassirostris commented:
@aknuds1 

> Should it be exec chroot --userspec=elasticsearch:elasticsearch / ./bin/es-docker?

That's my understanding, yes. Does it work for you?

> Could another alternative be to include the gosu executable in the Docker image?

I like it even less. If chroot is not an option, I'm OK with the current approach. Maybe @coffeepac will have some input here

aknuds1 commented:
@crassirostris I haven't tried the chroot approach yet, but I will. Out of curiosity though, why is including the gosu binary such a bad alternative?

crassirostris commented:
@aknuds1 OK, thanks!

> Out of curiosity though, why is including the gosu binary such a bad alternative?

Where will this binary be stored? How to update it? How to build it under a different OS/architecture? It's just easier not to solve these problems

aknuds1 commented:
@crassirostris I tried chroot; Elasticsearch fails because it needs to be run from within /usr/share/elasticsearch apparently. What should I do, create another wrapper script to be executed by chroot which changes the directory to /usr/share/elasticsearch?

aknuds1 commented:
@crassirostris I think I got the chroot approach working, by having chroot execute another script which changes the directory to /usr/share/elasticsearch before executing Elasticsearch itself. Should we go with this approach?

aknuds1 commented:
@crassirostris @coffeepac After playing around a bit, I decided to switch back to basing access to Kibana around `kubectl proxy`. The benefit is that I don't have to change the documentation, _and_ that one will go via the Kibana service instead of individual pods (like `kubectl port-forward`) forces you to. When forwarding to ports of individual pods you have to know the pod name, and it only works so long as the pod is alive. Additionally, if I want to change this behaviour it should be with a different PR.

Since I'm using the original Kibana image, I am setting the officially supported environment variable `SERVER_BASEPATH` instead of `KIBANA_BASE_URL`.

coffeepac commented:
I also prefer the non gosu approach for the reasons that @crassirostris stated, management overhead basically.  Having an extra wrapper script layer is a bit confusing but it is traceable which is important.

coffeepac commented:
@aknuds1 for the `Release Note` can you also include that you are switching to the upstream elasticsearch image?  That is, in my opinion, a worthy piece to call out.

aknuds1 commented:
@coffeepac How should I format this in the release note, write bullet points?

Kibana requires you to log in with the username/password.

coffeepac commented:
@aknuds1 bullet points is fine.  should only need the two:  upgrade to 5.5, switched to using official elastic image.

aknuds1 commented:
@coffeepac I edited the release note to note also that Elasticsearch and Kibana images correspond to the official ones (with customizations for the former)

aknuds1 commented:
@coffeepac Apparently, it's considered [bad practice to run Elasticsearch as root even under Docker](https://discuss.elastic.co/t/why-is-it-elasticsearch-is-not-allowed-to-run-as-root/60413).

coffeepac commented:
@aknuds1 thanks for that find.

coffeepac commented:
@aknuds1 one last thing and it should be good to go and my apologies for not seeing this earlier.  Can you add a note to the README that this now requires a default storage cluster be installed to function?  The volume claim template in the stateful set requires it.

crassirostris commented:
@aknuds1 @coffeepac I think it would also make sense to add a note to the README that you can (and probably should) change the default size and maybe even the storage type

crassirostris commented:
@aknuds1 I checked out the code and played with it for some time, couple of notes:

1. Changing users

Turns out the most stupid approach actually works (in `run.sh` file):

```
chown -R elasticsearch:elasticsearch /data

cd /usr/share/elasticsearch
./bin/elasticsearch_logging_discovery >> ./config/elasticsearch.yml
su elasticsearch -c ./bin/es-docker
```

2. `vm.max_map_count`

You removed [this line](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/fluentd-elasticsearch/es-image/run.sh#L17), and did a right thing, it indeed doesn't work. Let's fix this problem: write an init container which will run in a privileged mode and will run `sysctl -w vm.max_map_count=262144` if the value is less than `262144`.

3. chowning the volume

Instead of running as root and chowning a volume, you can have a following thing inside the podspec:

```
securityContext:
  fsGroup: 1000
```

That may need a PodSecurityPolicy, I think it supports narrowing down to a particular service account

aknuds1 commented:
@coffeepac @crassirostris I rewrote the fluentd-elasticearch README, please have a look.

aknuds1 commented:
@crassirostris About using `su` vs `gosu`, I'm not sure if there might be issues just using `su`. Isn't gosu written to handle problems with using su in a Docker/Kubernetes context?

crassirostris commented:
@aknuds1 

> About using su vs gosu, I'm not sure if there might be issues just using su. Isn't gosu written to handle problems with using su in a Docker/Kubernetes context?

The problem that's described in the [gosu README] is TTY and signal-forwarding. TTY is not a problem in K8s, and signal-forwarding just has to be checked. Could you please run the deployment and then send a signal to the root process in the ES container (that would be su I guess)?

coffeepac commented:
@aknuds1 README update looks good.  thank you.

aknuds1 commented:
@crassirostris 1. I've switched su for chroot
2. How do we do this?
3/4. Should we implement this and switch out su?

aknuds1 commented:
@crassirostris How do I test sending a signal to the root process?

crassirostris commented:
@aknuds1 Thanks!

> 2. How do we do this?

- https://kubernetes.io/docs/concepts/workloads/pods/init-containers/
- https://kubernetes.io/docs/tasks/configure-pod-container/security-context#set-the-security-context-for-a-container to set a privileged bit for this container

> 3/4. Should we implement this and switch out su?

Come to this about it, let's be more incremental, let's leave su for now. The main task is to make e2e tests work again, ES is broken now

> @crassirostris How do I test sending a signal to the root process?

From the host `ps aux | grep su` and e.g. `kill $PID`. Then see if the container was terminated and terminated gracefully (ES logs indicate graceful termination)

aknuds1 commented:
@crassirostris I added a privileged mode `initContainer` that sets `vm.max_map_count` as you asked.

crassirostris commented:
@aknuds1 I'm sorry, I was practically unavailable all this week

I'll review it over the weekend and hopefully we'll merge it by the EOD on Monday

crassirostris commented:
@coffeepac Are you OK with the PR?

coffeepac commented:
@crassirostris yep.  this is fine with me.  feel free to kick it along when you're satisfied.

crassirostris commented:
@aknuds1 Sorry, I didn't have time in the end

When I build the image, I get the following error:

```
Step 13 : COPY chrooted-run.sh bin/
lstat chrooted-run.sh: no such file or directory
```

aknuds1 commented:
@crassirostris Sorry about that, I will try again.

aknuds1 commented:
@crassirostris I fixed the build issue, and upgraded Elasticsearch and Kibana to 5.5.1. Elasticsearch looks fine, but I haven't yet been able to access Kibana as it takes some time to warm up

aknuds1 commented:
@crassirostris I implemented your suggestion of using a straight Alpine image for the initContainer. Have a look at the [change](https://github.com/kubernetes/kubernetes/pull/48722/files#diff-f5cffcbe9bbe0cc1bf6f8465719457b8R51) please.

crassirostris commented:
@aknuds1 LGTM, thanks a lot for your work!

Couple of last nits:

- Change the ES version in the release notes
- Add a paragraph about an init container & suggest to remove it if a system has a higher `vm.max_map_count` value

crassirostris commented:
Oh, and please squash commits

fhemberger commented on cluster/addons/fluentd-elasticsearch/es-image/Makefile:
```diff
@@ -16,7 +16,7 @@
 # The current value of the tag to be used for building and
 # pushing an image to gcr.io
-TAG = v5.4.0-1
+TAG = v5.5.1
```
@aknuds1 How about using the PREFIX, NAME, TAG notation as in `cluster/addons/fluentd-elasticsearch/kibana-image/Makefile` for consistency?

crassirostris replied:
I suggest merging this PR as it is right now, because it blocks some e2e tests and making a refactoring in a separate PR

fhemberger replied:
Fine for me, just discovered it as I'm eagerly waiting for this to land and I started looking into the Makefile. 😁

crassirostris commented on cluster/addons/fluentd-elasticsearch/README.md:
```diff
@@ -12,6 +11,13 @@ Elasticsearch is deployed as a
+[init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) executing
+before Elasticsearch containers themselves, in order to ensure that the kernel state variable
+`vm.max_map_count` is at least 262144, since this is a requirement of Elasticsearch.
+You may remove the init container if you know that your host filesystem meets this requirement.
```
It's not about filesystem, but OS

aknuds1 replied:
@crassirostris Ah, thank you! My bad.

aknuds1 commented:
@crassirostris Both nits handled. Please review my release notes as I added a couple.

crassirostris commented:
/lgtm

Thanks a lot for doing the work! One nit towards the release notes:

> Require authenticating towards Elasticsearch, as is the default

I don't understand what it means :) Last part at least, "as is the default". Could you please rephrase it?

crassirostris commented:
/approve no-issue

aknuds1 commented:
@crassirostris I revised my release notes, trying to make it more clear what I mean about requiring Elasticsearch authentication.

crassirostris commented:
@aknuds1 SGTM, thanks again! Sorry in advance for those test flakes =__=

sebglon commented:
Image are not pushed on not granted for public?

crassirostris commented:
@sebglon Was not pushed, sorry, Just fixed it

aknuds1 commented:
Thanks very much @crassirostris @coffeepac for helping out and getting this merged!

crassirostris commented:
I'm rolling back the PD, because it's leaking resources in the e2e tests

aknuds1 commented:
@crassirostris What do you mean by PD?

crassirostris commented:
@aknuds1 "Persistent disk", sorry, GCP terminology :) I mean PVC in the elasticsearch-logging

aknuds1 commented:
@crassirostris What's the solution then? Not having persistentVolumeClaimTemplate in the Elasticsearch StatefulSet (because of test issues)?

crassirostris commented:
@aknuds1 Yup, and document it better. Ideally, I'm thinking about splitting fluentd to a separate repo and having test yamls in this repo and an example with PVC in another

aknuds1 commented:
@crassirostris So the idea is to let the user define persistentVolumeClaimTemplate instead? You know best in any case.

crassirostris commented:
@aknuds1 The idea is to replace persistentVolumeClaimTemplate with emptyDir volume in this repo and add a note to the readme, stating that when adapting this example for your own need, you have to replace it with a PVC. Sounds good?

aknuds1 commented:
@crassirostris That's what I thought you would do. Sounds like the best solution everything taken into consideration!
